shared/src/ledger/storage/mod.rs

```diff
@@ -413,18 +516,158 @@ pub mod testing {
             let block = BlockStorage {
                 tree,
                 hash: BlockHash::default(),
-                height: BlockHeight(0),
+                height: BlockHeight::default(),
+                epoch: Epoch::default(),
                 subspaces,
             };
             Self {
                 db: MockDB::default(),
                 chain_id,
                 block,
-                current_height: BlockHeight(0),
+                current_height: BlockHeight::default(),
+                current_epoch: Epoch::default(),
+                next_epoch_min_start_height: BlockHeight::default(),
+                next_epoch_min_start_time: DateTimeUtc::now(),
                 address_gen: EstablishedAddressGen::new(
                     "Test address generator seed",
                 ),
             }
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use chrono::{TimeZone, Utc};
+    use proptest::prelude::*;
+
+    use super::testing::*;
+    use super::*;
+    use crate::ledger::parameters::Parameters;
+    use crate::types::time::{self, Duration};
+
+    prop_compose! {
+        /// Setup test input data with arbitrary epoch duration, epoch start
+        /// height and time, and a block height and time that are greater than
+        /// the epoch start height and time, and the change to be applied to
+        /// the epoch duration parameters.
+        fn arb_and_epoch_duration_start_and_block()
+        (
+            start_height in 0..1000_u64,
+            start_time in 0..10000_i64,
+            min_num_of_blocks in 1..10_u64,
+            min_duration in 1..100_i64,
+        )
+        (
+            min_num_of_blocks in Just(min_num_of_blocks),
+            min_duration in Just(min_duration),
+            start_height in Just(start_height),
+            start_time in Just(start_time),
+            block_height in start_height + 1..(start_height + 2 * min_num_of_blocks as u64),
+            block_time in start_time + 1..(start_time + 2 * min_duration),
+            min_blocks_delta in -(min_num_of_blocks as i64 - 1)..5,
```

Comment:
I assume this `-`  in front of `(min_num_of_blocks...)` is not a mistake, but it is not clear what it should mean

Reply:
I've added some comments here: `Delta will be applied on the min_num_of_blocks parameter`. The function generates a random number from this range, the left bound is set to a value so that after the delta is applied, the `min_num_of_blocks` doesn't go below 1 (`min_num_of_blocks - (min_num_of_blocks - 1) == 1`

Reply:
:+1: